src/TypeScriptGenerator.ts

```diff
@@ -144,44 +139,51 @@ function selectionsToAST(
       true
     );
     types.push([otherPropWithComment]);
-  } else {
-    let selectionMap = selectionsToMap(Array.from(baseFields.values()));
-    for (const concreteType in byConcreteType) {
-      selectionMap = mergeSelections(
-        selectionMap,
-        selectionsToMap(
-          byConcreteType[concreteType].map(sel => ({
-            ...sel,
-            conditional: true
-          }))
-        )
-      );
-    }
-    const selectionMapValues = groupRefs(Array.from(selectionMap.values())).map(
-      sel =>
-        isTypenameSelection(sel) && sel.concreteType
-          ? makeProp({ ...sel, conditional: false }, state, sel.concreteType)
-          : makeProp(sel, state)
-    );
-    types.push(selectionMapValues);
   }
 
-  return ts.createUnionTypeNode(
-    types.map(props => {
-      if (refTypeName) {
-        props.push(
-          readOnlyObjectTypeProperty(
-            REF_TYPE,
-            ts.createTypeReferenceNode(
-              ts.createIdentifier(refTypeName),
-              undefined
-            )
-          )
-        );
-      }
-      return exactObjectTypeAnnotation(props);
-    })
+  let selectionMap = selectionsToMap(Array.from(baseFields.values()));
+  for (const concreteType in byConcreteType) {
+    selectionMap = mergeSelections(
+      selectionMap,
+      selectionsToMap(
+        byConcreteType[concreteType].map(sel => ({
+          ...sel,
+          conditional: true
+        }))
+      )
+    );
+  }
+  const baseProps: ts.PropertySignature[] = groupRefs(
+    Array.from(selectionMap.values())
+  ).map(
+    sel =>
+      isTypenameSelection(sel) && sel.concreteType
+        ? makeProp({ ...sel, conditional: false }, state, sel.concreteType)
+        : makeProp(sel, state)
   );
+
+  if (refTypeName) {
+    baseProps.push(
+      readOnlyObjectTypeProperty(
+        REF_TYPE,
+        ts.createTypeReferenceNode(ts.createIdentifier(refTypeName), undefined)
+      )
+    );
+  }
+
+  if (types.length > 0) {
+    const unionType = ts.createUnionTypeNode(
+      types.map(props => {
+        return exactObjectTypeAnnotation(props);
+      })
+    );
+    return ts.createIntersectionTypeNode([
+      exactObjectTypeAnnotation(baseProps),
+      unionType
+    ]);
+  } else {
+    return exactObjectTypeAnnotation(baseProps);
+  }
```

Comment:
This is going to generate a lot of noise for people with large sets of artefacts, but it should be fine in functionality 👍